Plash/URLPlaceholder.swift

```diff
@@ -0,0 +1,54 @@
+//
+//  URLPlaceholder.swift
+//  Plash
+//
+//  Created by Jonathan Cole on 8/31/20.
+//  Copyright © 2020 Sindre Sorhus. All rights reserved.
+//
+
+import Foundation
+
+/**
+Defines a placeholder to be substituted out for a string in a URL.
+*/
+struct URLPlaceholder {
+	/// The string that will be replaced by the output of `transform()`.
+	var token: String
+	/// The string this function returns will replace any occurrences of `token`.
+	var transform: () -> String
+
+	/**
+	Replace any instances of `token` with the value returned by `transform()`.
+	*/
+	func appliedTo(url: URL) throws -> URL {
+		guard let token = self.token.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) else {
+			throw URLPlaceholderError.failedToEncodeToken(self.token)
+		}
+		let urlString = url.absoluteString
+			.replacingOccurrences(of: token, with: self.transform())
+		guard let newURL = URL(string: urlString) else {
+			throw URLPlaceholderError.invalidURLAfterSubstitution(urlString)
+		}
+		return newURL
+	}
+}
+
+enum URLPlaceholderError: Error, LocalizedError {
+	case failedToEncodeToken(String)
+	case invalidURLAfterSubstitution(String)
+
+	var errorDescription: String? {
+		switch self {
+		case .failedToEncodeToken(let token):
+			return "Failed to encode token \"\(token)\""
+		case .invalidURLAfterSubstitution(let urlString):
+			return "New URL was not valid after substituting placeholders. URL string is \"\(urlString)\""
+		}
+	}
+}
+
+extension URL {
+	func replacingPlaceholder(_ placeholder: URLPlaceholder) throws -> URL {
```

Comment:
Why not just have this method accept a `token` and `transform` directly as arguments? It's not clear what benefit using a struct here gives.

Reply:
I agree that it can be simplified. If we keep them as structs we could potentially do something like serializing a list of them as buttons to facilitate easier insertion for users. What's your preference!

Reply:
I prefer focusing on the current case. It can be refactored in the future if there are other needs. But I try not to design for imaginary use-cases. So I prefer to drop the struct.

Reply:
That makes sense. Will do.